File: provider/pkg/provider/ready.go
package provider

import (
	"github.com/pulumi/pulumi-command/sdk/go/command/remote"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// The set of arguments for creating a Ready component resource.
type ReadyArgs struct {
	Connection    remote.ConnectionInput `pulumi:"connection" provider:"secret"`
	CustomTimeout string                 `pulumi:"customTimeout"`
}

// The Ready component resource.
type Ready struct {
	pulumi.ResourceState
}

// NewReady waits for an SSH connection to be active.
func NewReady(ctx *pulumi.Context,
	name string, args *ReadyArgs, opts ...pulumi.ResourceOption) (*Ready, error) {
	if args == nil {
		args = &ReadyArgs{}
	}

	// Set default values if not provided
	if args.CustomTimeout == "" {
		args.CustomTimeout = "10m" // Default timeout
	}

	ready := &Ready{}
	err := ctx.RegisterComponentResource("svmkit:index:Ready", name, ready, opts...)
	if err != nil {
		return nil, err
	}

	_, err = remote.NewCommand(ctx, "isReady", &remote.CommandArgs{
		Create:     pulumi.String("echo connected"),
		Connection: args.Connection,
	}, pulumi.Parent(ready), pulumi.Timeouts(&pulumi.CustomTimeouts{Create: args.CustomTimeout}))
	if err != nil {
		return nil, err
	}

	if err := ctx.RegisterResourceOutputs(ready, pulumi.Map{}); err != nil {
		return nil, err
	}

	return ready, nil
}

File: provider/cmd/pulumi-resource-svmkit/ready.ts
import * as pulumi from "@pulumi/pulumi";
import * as provider from "@pulumi/pulumi/provider";
import * as command from "@pulumi/command";
import { naming } from "./name";

export interface ReadyArgs {
    timeout?: string;
    connection: command.types.input.remote.ConnectionArgs;
}

export class Ready extends pulumi.ComponentResource {
    constructor(
        name: string,
        args: ReadyArgs,
        opts?: pulumi.ComponentResourceOptions,
    ) {
        super("svmkit:index:Ready", name, args, opts);
        const parent = this;

        const tag = naming(name);

        const { connection } = args;

        const timeout = args.timeout ?? "10m";

        new command.remote.Command(
            tag("is", "machine", "running"),
            {
                connection: { ...connection, dialErrorLimit: -1 },
                create: "echo connected",
            },
            {
                parent,
                customTimeouts: { create: timeout },
            },
        );
    }
}

export async function constructReady(
    name: string,
    inputs: pulumi.Inputs,
    options: pulumi.ComponentResourceOptions,
): Promise<provider.ConstructResult> {
    const ready = new Ready(name, inputs as ReadyArgs, options);

    return {
        urn: ready.urn,
        state: {},
    };
}
